rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+
+```rust
+impl Db {
+    /// Scan a range of keys with the provided options.
+    ///
+    /// returns a `DbIterator`
+    pub async fn scan_with_options(
+        &self,
+        range: &DbRange,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+pub struct ScanOptions {
+    /// The read commit level for read operations
+    pub read_level: ReadLevel,
+    /// The number of bytes to read ahead
+    pub read_ahead_size: usize,
+    /// Whether or not fetched blocks should be cached
+    pub cache_blocks: bool
+}
+```
+
+## Implementation
+
+A scan must iterate every table which may have keys within the query range.
+This includes tables in memory which which have not been flushed and L0/SR tables 
+which may need to be loaded from object storage. Internally, we will rely on 
+iterators which sort and merge these tables. The latest records will have precedence following
+the same rules that a normal `get` query follows.
+
+The existing `MergeIterator` used for compaction provides much of the necessary
+foundation to unify multiple iterators. Additionally, `MemTableIterator` and 
+`SstIterator` provide ready-to-use capabilities to scan tables which may have keys
+within the range. 
+
+**Query Isolation**: When using the `Uncommitted` isolation, unflushed
+records recorded in WALs will be used as a source for the key range. Otherwise,
+only committed records in memtables, as well as L0 and SR tables will be used.
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the query began scanning its range of records. For example,
+a query may include only a subset of records that were inserted in the same 
+`WriteBatch`. In other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-????](https://github.com/slatedb/slatedb/pull/278) in order to prevent the cleanup
+of durable state referenced by the query. When a range scan is created, it will create
+an internal reference to the current checkpoint. This reference will prevent cleanup of
+the checkpoint state until the scan is complete. 
+
+Checkpoints created by `Readers` have an expiry time, which is only refreshed as long
+as it remains the current checkpoint. If a `DbIterator` instance remains active long 
+enough for its referenced checkpoint to expire, the `Reader` will proactively invalidate 
```

Comment:
Do we really need to invalidate the iterator when the checkpoint expires? Alternatively, we could do this when a call to `next` fails because it did a `get` and the SST object was not found.

Reply:
Agree that works too. I guess the main thing I was hoping for was a clear error, but we could wrap the not found error as well.